source/common/http/filter/gzip_filter.cc

```diff
@@ -145,65 +146,127 @@ FilterDataStatus GzipFilter::encodeData(Buffer::Instance& data, bool end_stream)
   return Http::FilterDataStatus::StopIterationNoBuffer;
 }
 
-bool GzipFilter::isMinimumContentLength(HeaderMap& headers) const {
-  const Http::HeaderEntry* content_length = headers.ContentLength();
-  if (content_length) {
-    uint64_t length{};
-    StringUtil::atoul(content_length->value().c_str(), length);
-    return length >= config_->minimumLength();
+bool GzipFilter::isAcceptEncodingAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* accept_encoding = headers.AcceptEncoding();
+  // TODO(gsagula): Since gzip is the only available content-encoding in Envoy at the moment,
+  // order/priority of preferred server encodings is disregarded (RFC2616-14.3). Replace this
+  // with a data structure that parses Accept-Encoding values and allows fast lookup of
+  // key/priority. Also, this should be part of some utility library.
+  if (accept_encoding) {
+    bool is_identity{true};  // true if does not exist or is followed by q=0.
+    bool is_wildcard{false}; // true if found and not followed by `q=0`.
+    bool is_gzip{false};     // true if exists and not followed by `q=0`.
+
+    auto tokens = StringViewUtil::splitToken(headers.AcceptEncoding()->value().c_str(), ";,");
+    for (auto it = tokens.begin(); it != tokens.end(); it++) {
```

Comment:
Thread sanitizer is throwing heap-use-after-freer here. I will need to look into that later.

Reply:
I need to wrap my arms around the memory lifecycle here with the accept-encoding lookup in this library, but for now why don't you capture the accept-encoding return in a temp.  You can also use a range loop for brevity.

  LowerCaseString accept_encoding = StringViewUtil::splitToken(headers.AcceptEncoding()->value();
  for (auto token : StringViewUtil::splitToken(accept_encoding.c_str(), ";,") { ...

don't forget to trim any the whitespace around the tokens.